### ISSUE TITLE
Newbie question: can it execute Gherkin feature files?

### ISSUE THREAD

**Issue by davix** [adamluzsi/testcase#5]
Thanks

**Comment by adamluzsi** [comment]
Hello @davix, 

Thank you for your question!
No, this library is not meant to parse Gherkin feature files.

While Gherkin-based feature files are great to build a bridge between business experts and technical experts, this is not the goal of this project.

The project's long-term goal is to make unit testing [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself), modular, and maintainable.
It is only meant to be used by technical experts.

It allows developers to convert any testcase unit test into a fuzz test (fuzzing).
It even allows using unit tests as integration tests if the testcase project's convention was followed.

While Gherkin could be a great tool to discuss and define high-level E2E expectations with business experts, they are often far from being easy to maintain, so they are not ideal for unit testing.

Was I able to answer your question and also explain the why?

**Comment by davix** [comment]
Thanks @adamluzsi for the explanation.
I was trying to find better alternatives for godog.
I thought BDD is for business experts as well, otherwise technical experts can choose any way they like in UT/integration/e2e.
I haven't read through all content in readme, will try to understand the project goal.

**Comment by davix** [comment]
Or maybe I can use godog to parse feature files and map the GIVEN/WHEN/THEN step implementation to testcase.Spec's .When/.Then?

**Comment by adamluzsi** [comment]
[tdd-wiki-link]: https://en.wikipedia.org/wiki/Test-driven_development
[bdd-wiki-link]: https://en.wikipedia.org/wiki/Behavior-driven_development
[cucumber-link]: https://cucumber.io/

Hello @davix,

First of all, I'm aware that I lack awefully lot of context about your project, team and overall setup.
Due to time constraints, I will assume some of them, and try to give a bit of my insight.
To understand what is the right tool for your need, I would like to explain the concept of BDD first.

## So, what is TDD/BDD testing?

TL;DR
> **Testing** is commonly a core practice in development.
  **Manual testing** and testing in general aim to confirm the system's expected behavior.
  **Automatic tests** continuously test these system behavior expectations without consuming human time.
  Additionally, it enables developers to work without memorizing these past system expectations, 
  **TDD** is the practice of writing tests for future code and ensures simplicity and clarity.
  Through this, no new code should be present in the system without automatic tests.
  **BDD** specifies **what** to test, and **discipline** the user to focus on system behavior rather than implementation details.

**Manual Testing** is a methodology where the code's expected behavior is being tested.

**Automatic testing** is based on top of this by describing the practice to create tests that can run automatically.
This allows the developers to forget features when they are not in scope, and focus on the new changes.

**[TDD][tdd-wiki-link]** is extending this by defining **when** that test should be written.
It gives a set of practices such as following the RED/GREEN/REFACTOR pattern,
where a failing automatic test must justify the act of writing a production test.

On top of [TDD][tdd-wiki-link], there is [BDD][bdd-wiki-link] which extends further the testing practices by requesting **discipline** on **what** you test in your automated tests.
In a simplified summary, [BDD][tdd-wiki-link] principles discipline the test writer to focus on testing the **behavior** instead of the implementation.
By testing behavior over implementation, the ability to refactor the production code base becomes significantly easier.
Also if testing behavior over implementation turns out challenging, it often indicates architecture design smells in the project.

Since this testing approach is not specific to any of the architecture layers, you are not limited to only do this on an E2E level, but instead, you can follow the classic testing pyramid concept and keep the focus on the unit tests.

## What is not BDD?

The usage of `Given`, `When`, `And`, `Then` keywords doesn't automatically make tests fit into BDD 
These keywords are often used for describing a scenario in a BDD testing style,
but they are syntax sugar rather than the principles of BDD.

## when/what?

Generally, I would stick with the testing pyramid focusing on quality unit tests and only use Jira's Xray feature as a tool to define **Acceptance Criteria** rather than bound the application's testing suite stability to an external service, and force a very high-level language as the defacto testing approach for the project. Otherwise, your team will eventually start to feel the downside of switching the direction of the testing pyramid. The project's testing suite will become sluggish, the errors from the feedback will be too high level, and finding the root cause will become a time-wasting activity, not the full testing suite will be executed eventually and CI/CD pipelines will often start to fail because the skipped tests that `shouldn't be affected by the change`, and so on.

For defining really high level **Acceptance Criteria** with **Jira Xray** in the form of Gherkin, that shouldn't be an issue, but a welcome addition to the project's testing suite.
For that purpose, I believe any tool that is able to run Gherkin feature files based on E2E tests should be fine, such as GoDog or Gauge. Technically you don't even have to integrate that into the individual project, but could have a separate repository that focuses on E2E testing, and whenever a project's pipeline passes, it can trigger a next step where these tests being confirmed.

![image](https://user-images.githubusercontent.com/2897081/130786676-5ff489c6-4f54-4ad6-b1e5-84e502444c0e.png)

## ps

What do you think?

Believe me, I would be more than honored if your team would decide to use `testcase`!
I know I'm making this less likely to happen by the above explanation, but I truly believe that `testcase` is either not the right tool for your issue.

Have you considered if Jira's Xray could be a separate small QA-focused project on its own within your organization?
That way it would be less likely to ruin stability and the testing pyramid balance in your individual projects.

Cheers,
Adam L.

**Comment by davix** [comment]
> From a domain expert point of view, 80% of the unit test is noise to them, as they most of the time small application-level implementation detail.

I didn't mean to implement UT with Gherkin or replace any UT framework, I meant Gherkin may be implemented with UT per requirement. (some implemented at e2e level, some at UT level. Product owner doesn't need to know)

Interesting, I happen to see book "BDD in Action" talks about it as well.
<img width="771" alt="01_09" src="https://user-images.githubusercontent.com/2516468/131768216-921f8033-d7e7-4ebc-aa20-6aa74292e728.png">

Thank you @adamluzsi for so much clarification.

**Comment by adamluzsi** [comment]
@davix I'm happy if I was any use for your research! Sorry if I might sounded a bit sharp in my tone. I misinterpreted what you meant by "gherkin implemented by unit test", and it brought back memories when my team learned using gherkin through the hard way. I wanted to save a rough ride for you. But using Gherkin for high level acceptance criterion was great experience. If your team do that, I believe it will be only beneficial for the team.

If you have further questions regarding testing, don't hesitate to ask. Otherwise if your question is answered, please close the ticket.

Cheers.
@adamluzsi

**Comment by adamluzsi** [comment]
Also, @davix don't forget to leave a star on the project if you like it. ;)

**Comment by davix** [comment]
Star for sure. My original question was answered, thanks!
